composer/callbacks/checkpoint_saver.py

```diff
@@ -431,19 +458,42 @@ def _save_checkpoint(self, state: State, logger: Logger):
                 # create and upload a symlink file
                 with tempfile.TemporaryDirectory() as tmpdir:
                     symlink_filename = os.path.join(tmpdir, 'latest.symlink')
-                    create_symlink_file(remote_file_name, symlink_filename)
+                    # Sharded checkpoints for torch >2.0 use directories not files for load_paths
+                    if state.fsdp_elastic_sharded_enabled:
+                        src_path = str(pathlib.Path(remote_file_name).parent)
+                    else:
+                        src_path = remote_file_name
+                    log.debug(f'Creating symlink file {symlink_filename} -> {src_path}')
+                    create_symlink_file(src_path, symlink_filename)
                     logger.upload_file(
                         remote_file_name=symlink_name,
                         file_path=symlink_filename,
                         overwrite=True,
                     )
 
-        self.saved_checkpoints.append(saved_path)
+        if dist.get_global_rank() == 0 and state.fsdp_elastic_sharded_enabled:
+            assert metadata_local_file_path is not None
+            self.saved_checkpoints.append((saved_path, metadata_local_file_path))
+        else:
+            self.saved_checkpoints.append(saved_path)
 
         if self.num_checkpoints_to_keep >= 0:
-            self._rotate_checkpoints()
+            self._rotate_checkpoints(sharding_enabled=state.fsdp_sharded_state_dict_enabled)
+
+    def _rotate_checkpoints(self, sharding_enabled: bool = False):
```

Comment:
Is this rotation stuff tested with sharded checkpoints anywhere?

Reply:
yes test_cleanup_sharded_checkpoints